File: sdk/typescript/runtime/src/tool-generator.ts
/**
 * LAVS Tool Generator
 *
 * Automatically generates Claude SDK tool definitions from LAVS manifests.
 * This allows AI agents to call LAVS endpoints as tools.
 */

import { LAVSManifest, Endpoint } from './types';
import { ManifestLoader } from './loader';
import { ScriptExecutor } from './script-executor';
import { FunctionExecutor } from './function-executor';
import { LAVSValidator } from './validator';
import { PermissionChecker } from './permission-checker';
import { ScriptHandler, FunctionHandler, HTTPHandler, MCPHandler, ExecutionContext } from './types';
import { HttpExecutor } from './http-executor';
import { McpExecutor } from './mcp-executor';
import { debug } from './logger';
import path from 'path';
import http from 'http';

/**
 * The default port the LAVS host listens on. Can be overridden via the
 * LAVS_HOST_PORT environment variable — this lets `lavs call` notify a host
 * running on a non-default port (e.g. when `lavs view --port 9000` is used,
 * or when the daemon plist configures a custom port).
 */
function getGlobalHostPort(): number {
  const env = process.env.LAVS_HOST_PORT;
  if (env) {
    const n = parseInt(env, 10);
    if (!Number.isNaN(n) && n > 0) return n;
  }
  return 7842;
}

/**
 * Fire-and-forget: notify the global LAVS host that a mutation or UI
 * command happened so connected view iframes can react via SSE.
 * `input` is only set for `notify` endpoints (the command arguments).
 */
function notifyGlobalHost(bundleName: string, endpointId: string, data: unknown, input?: unknown): void {
  try {
    const body = JSON.stringify(input !== undefined ? { data, input, kind: 'ui_command' } : { data });
    const req = http.request({
      hostname: '127.0.0.1',
      port: getGlobalHostPort(),
      path: `/api/notify/${encodeURIComponent(bundleName)}/${encodeURIComponent(endpointId)}`,
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(body) },
    });
    req.on('error', () => { /* host not running — silently ignore */ });
    req.write(body);
    req.end();
  } catch {
    // Ignore errors — host notification is best-effort
  }
}

/**
 * Claude SDK tool definition
 */
export interface ClaudeTool {
  name: string;
  description: string;
  input_schema: {
    type: 'object';
    properties: Record<string, any>;
    required?: string[];
  };
}

/**
 * Tool execution function
 */
export type ToolExecutor = (params: any) => Promise<any>;

/**
 * Generated tool with executor
 */
export interface GeneratedTool {
  tool: ClaudeTool;
  execute: ToolExecutor;
  /** Endpoint method type — used by MCP server to format output and decide view refresh */
  method: 'query' | 'mutation' | 'notify';
}

/**
 * Generate tools from LAVS manifest
 */
export class LAVSToolGenerator {
  /**
   * Generate tools for an agent
   * @param agentId - Agent ID
   * @param agentDir - Agent directory path
   * @param projectPath - Optional project path for data isolation
   * @returns Array of generated tools
   */
  async generateTools(
    agentId: string,
    agentDir: string,
    projectPath?: string
  ): Promise<GeneratedTool[]> {
    try {
      // Load manifest
      const lavsPath = path.join(agentDir, 'lavs.json');
      const loader = new ManifestLoader();
      const manifest = await loader.load(lavsPath);

      // Generate tool for each endpoint
      const tools: GeneratedTool[] = [];

      for (const endpoint of manifest.endpoints) {
        // Only generate tools for query and mutation endpoints
        // Subscriptions don't make sense as tools
        if (endpoint.method === 'subscription') {
          continue;
        }

        const tool = this.generateToolForEndpoint(endpoint, manifest, agentId, agentDir, projectPath);
        tools.push(tool);
      }

      debug(`[LAVS] Generated ${tools.length} tools for agent ${agentId}`);
      return tools;
    } catch (error: unknown) {
      // If no lavs.json, that's OK - just return empty array
      const message = error instanceof Error ? error.message : '';
      if (message.includes('not found')) {
        return [];
      }
      throw error;
    }
  }

  /**
   * Generate tool definition and executor for an endpoint
   */
  private generateToolForEndpoint(
    endpoint: Endpoint,
    manifest: LAVSManifest,
    agentId: string,
    agentDir: string,
    projectPath?: string
  ): GeneratedTool {
    // Generate tool name (prefix with lavs_ to avoid conflicts)
    const toolName = `lavs_${endpoint.id}`;

    // Generate tool description
    const description = endpoint.description || `Call ${endpoint.id} endpoint from ${manifest.name}`;

    // Generate input schema from endpoint schema
    const inputSchema = endpoint.schema?.input || {
      type: 'object',
      properties: {},
    };

    // Ensure it's an object schema
    if (inputSchema.type !== 'object') {
      throw new Error(`Endpoint ${endpoint.id} must have object input schema`);
    }

    const tool: ClaudeTool = {
      name: toolName,
      description,
      input_schema: {
        type: 'object',
        properties: inputSchema.properties || {},
        required: inputSchema.required || [],
      },
    };

    // Create shared instances for validation and permission checking
    const validator = new LAVSValidator();
    const permChecker = new PermissionChecker();
    const manifestTypes = manifest.types;

    // Create executor function
    const execute: ToolExecutor = async (params: any) => {
      debug(`[LAVS] Executing tool ${toolName} with params:`, params);

      // 1. Validate input against schema (pass manifest types for $ref resolution)
      validator.assertValidInput(endpoint, params, manifestTypes);

      // 2. Merge permissions
      const mergedPermissions = permChecker.mergePermissions(
        manifest.permissions,
        endpoint.permissions
      );

      // 3. Check permissions for script handlers
      if (endpoint.handler?.type === 'script') {
        permChecker.assertAllowed(
          endpoint.handler as ScriptHandler,
          mergedPermissions,
          agentDir
        );
      }

      // 4. Build execution context
      const context: ExecutionContext = {
        endpointId: endpoint.id,
        agentId,
        workdir: agentDir,
        permissions: mergedPermissions,
        env: projectPath ? {
          LAVS_PROJECT_PATH: projectPath,
        } : undefined,
      };

      // 5. Execute the handler. `notify` endpoints may omit the handler
      // entirely — a pure UI command has nothing to run server-side; the
      // broadcast below is the whole effect.
      let result: unknown;
      if (!endpoint.handler) {
        result = { ok: true };
      } else switch (endpoint.handler.type) {
        case 'script': {
          const executor = new ScriptExecutor();
          result = await executor.execute(
            endpoint.handler as ScriptHandler,
            params,
            context
          );
          break;
        }
        case 'function': {
          const funcExecutor = new FunctionExecutor();
          result = await funcExecutor.execute(
            endpoint.handler as FunctionHandler,
            params,
            context
          );
          break;
        }
        case 'http': {
          const httpExecutor = new HttpExecutor();
          result = await httpExecutor.execute(
            endpoint.handler as HTTPHandler,
            params,
            context
          );
          break;
        }
        case 'mcp': {
          const mcpExecutor = new McpExecutor();
          result = await mcpExecutor.execute(
            endpoint.handler as MCPHandler,
            params,
            context
          );
          break;
        }
        default: {
          const htype = (endpoint.handler as any).type ?? 'unknown';
          throw new Error(`Handler type '${htype}' is not yet supported in tool generation`);
        }
      }

      // 6. Notify global host after mutations / UI commands so views react
      // (fire-and-forget). Skip when called from within the host itself
      // (LAVS_HOST_CALLER=1) to avoid double-broadcasting — the host's
      // /api/call handler broadcasts directly.
      if ((endpoint.method === 'mutation' || endpoint.method === 'notify') && !process.env.LAVS_HOST_CALLER) {
        notifyGlobalHost(
          manifest.name,
          endpoint.id,
          result,
          endpoint.method === 'notify' ? params : undefined
        );
      }

      // 7. Validate output against schema (non-blocking: warn on mismatch, still return data)
      try {
        validator.assertValidOutput(endpoint, result, manifestTypes);
      } catch (validationError: any) {
        console.warn(`[LAVS] Output validation warning for ${toolName}: ${validationError.message}`);
      }

      return result;
    };

    return { tool, execute, method: endpoint.method as 'query' | 'mutation' | 'notify' };
  }

  /**
   * Check if agent has LAVS
   */
  async hasLAVS(agentDir: string): Promise<boolean> {
    try {
      const lavsPath = path.join(agentDir, 'lavs.json');
      const loader = new ManifestLoader();
      await loader.load(lavsPath);
      return true;
    } catch {
      return false;
    }
  }
}

File: sdk/python/lavs_runtime/tool_generator.py
"""
LAVS Tool Generator.

Generates Claude SDK / MCP tool definitions from LAVS manifests and
dispatches calls to the appropriate handler executor.

Mirrors the TypeScript LAVSToolGenerator class; currently supports
``script`` handler type end-to-end. ``http`` and ``mcp`` handler types
are declared but raise NotImplementedError until their executors are added.
"""

from __future__ import annotations

import json
import logging
import os
import urllib.parse
import urllib.request
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Literal

from lavs_runtime.loader import ManifestLoader
from lavs_runtime.permission_checker import PermissionChecker
from lavs_runtime.script_executor import ScriptExecutor
from lavs_runtime.validator import LAVSValidator
from lavs_types import (
    Endpoint,
    ExecutionContext,
    LAVSError,
    LAVSErrorCode,
    LAVSManifest,
)

logger = logging.getLogger(__name__)

_DEFAULT_HOST_PORT = 7842


def _get_host_port() -> int:
    env = os.environ.get("LAVS_HOST_PORT", "")
    try:
        n = int(env)
        if n > 0:
            return n
    except ValueError:
        pass
    return _DEFAULT_HOST_PORT


def _notify_host(bundle_name: str, endpoint_id: str, data: Any) -> None:
    """Fire-and-forget POST to the running LAVS host after a mutation."""
    try:
        port = _get_host_port()
        body = json.dumps({"data": data}).encode()
        url = (
            f"http://127.0.0.1:{port}/api/notify/"
            f"{urllib.parse.quote(bundle_name, safe='')}"
            f"/{urllib.parse.quote(endpoint_id, safe='')}"
        )
        req = urllib.request.Request(
            url,
            data=body,
            headers={"Content-Type": "application/json"},
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=0.5):
            pass
    except Exception:
        pass  # host not running — silently ignore


# ---------------------------------------------------------------------------
# Public types
# ---------------------------------------------------------------------------


@dataclass
class ClaudeTool:
    """Claude SDK / MCP tool definition."""

    name: str
    description: str
    input_schema: dict[str, Any] = field(
        default_factory=lambda: {"type": "object", "properties": {}},
    )

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "description": self.description,
            "input_schema": self.input_schema,
        }


ToolExecutor = Callable[[Any], Any]


@dataclass
class GeneratedTool:
    """A tool definition paired with its executor callable."""

    tool: ClaudeTool
    execute: ToolExecutor
    method: Literal["query", "mutation"]


# ---------------------------------------------------------------------------
# Generator
# ---------------------------------------------------------------------------


class LAVSToolGenerator:
    """
    Generate tool definitions from a LAVS manifest and dispatch calls.

    Usage::

        gen = LAVSToolGenerator()
        tools = gen.generate_tools(agent_id="myagent", agent_dir="/path/to/bundle")
        for gt in tools:
            result = gt.execute({"key": "value"})
    """

    def generate_tools(
        self,
        agent_id: str,
        agent_dir: str,
        project_path: str | None = None,
    ) -> list[GeneratedTool]:
        """
        Load the manifest in *agent_dir* and return one :class:`GeneratedTool`
        per non-subscription endpoint.

        Args:
            agent_id: Logical identifier for this agent instance.
            agent_dir: Directory containing ``lavs.json``.
            project_path: Optional data-isolation path forwarded as
                ``LAVS_PROJECT_PATH`` to script handlers.

        Returns:
            List of generated tools (empty if no ``lavs.json`` found).

        Raises:
            :class:`lavs_types.LAVSError`: On manifest parse / validation errors.
        """
        manifest_path = str(Path(agent_dir) / "lavs.json")
        loader = ManifestLoader()
        try:
            manifest = loader.load(manifest_path)
        except LAVSError as exc:
            if "not found" in str(exc).lower():
                return []
            raise

        tools: list[GeneratedTool] = []
        for endpoint in manifest.endpoints:
            if endpoint.method == "subscription":
                continue
            gt = self._build_tool(endpoint, manifest, agent_id, agent_dir, project_path)
            tools.append(gt)

        logger.debug("[LAVS] Generated %d tools for agent %s", len(tools), agent_id)
        return tools

    def has_lavs(self, agent_dir: str) -> bool:
        """Return True if *agent_dir* contains a valid ``lavs.json``."""
        try:
            manifest_path = str(Path(agent_dir) / "lavs.json")
            ManifestLoader().load(manifest_path)
            return True
        except Exception:
            return False

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _build_tool(
        self,
        endpoint: Endpoint,
        manifest: LAVSManifest,
        agent_id: str,
        agent_dir: str,
        project_path: str | None,
    ) -> GeneratedTool:
        tool_name = f"lavs_{endpoint.id}"
        description = endpoint.description or f"Call {endpoint.id} from {manifest.name}"

        raw_schema = {}
        if endpoint.endpoint_schema and endpoint.endpoint_schema.input:
            raw_schema = endpoint.endpoint_schema.input
            if isinstance(raw_schema, dict):
                pass
            else:
                raw_schema = raw_schema.model_dump(exclude_none=True)

        input_schema: dict[str, Any] = {
            "type": "object",
            "properties": raw_schema.get("properties", {}),
        }
        if "required" in raw_schema:
            input_schema["required"] = raw_schema["required"]

        claude_tool = ClaudeTool(
            name=tool_name,
            description=description,
            input_schema=input_schema,
        )

        validator = LAVSValidator()
        perm_checker = PermissionChecker()
        manifest_types = manifest.types

        def execute(params: Any) -> Any:
            # 1. Validate input
            validator.assert_valid_input(endpoint, params, manifest_types)

            # 2. Merge permissions
            merged = perm_checker.merge_permissions(manifest.permissions, endpoint.permissions)

            # 3. Build execution context
            context_env: dict[str, str] | None = (
                {"LAVS_PROJECT_PATH": project_path} if project_path else None
            )
            context = ExecutionContext(
                endpoint_id=endpoint.id,
                agent_id=agent_id,
                workdir=agent_dir,
                permissions=merged,
                env=context_env,
            )

            # 4. Execute handler
            handler = endpoint.handler

            result: Any
            if getattr(handler, "type", None) == "script":
                # Permission check for script handlers
                perm_checker.assert_allowed(handler, merged, agent_dir)  # type: ignore[arg-type]
                executor = ScriptExecutor()
                result = executor.execute(handler, params, context)  # type: ignore[arg-type]
            elif getattr(handler, "type", None) in ("http", "mcp", "function"):
                raise NotImplementedError(
                    f"Handler type '{handler.type}' is not yet supported "  # type: ignore[union-attr]
                    "in the Python runtime tool generator. "
                    "Use the TypeScript runtime for http/mcp/function handlers."
                )
            else:
                raise LAVSError(
                    LAVSErrorCode.HandlerError,
                    f"Unknown handler type: {getattr(handler, 'type', 'unknown')}",
                )

            # 5. Notify host after mutations (fire-and-forget)
            if endpoint.method == "mutation" and not os.environ.get("LAVS_HOST_CALLER"):
                _notify_host(manifest.name, endpoint.id, result)

            # 6. Validate output (advisory: warn, don't raise)
            try:
                validator.assert_valid_output(endpoint, result, manifest_types)
            except LAVSError as warn_exc:
                logger.warning("[LAVS] Output validation warning for %s: %s", tool_name, warn_exc)

            return result

        return GeneratedTool(
            tool=claude_tool,
            execute=execute,
            method=endpoint.method,  # type: ignore[arg-type]
        )
